### PR TITLE
[WIP] TF Trainer with TPUs

### LINKED ISSUE #5042: ❓ [TFTrainer] How to run on 8 TPU cores ?
# ❓ Questions & Help

I'm trying to run TFTrainer on 8 TPU cores, but I don't understand how to make it work.

I tried running my script with the flags `--tpu_num_cores 8 --per_device_train_batch_size 8`, expecting each core to handle a batch size of 1.

But when I print the shape of my inputs, I have `[8, x]` instead of `[1, x]`, which leads to memory error.

---

If I start the training with `--tpu_num_cores 8 --per_device_train_batch_size 1`, the shape of inputs is correct (`[1, x]`), but the number of optimization steps computed is not correct (If I have 8k samples, it says I have 8k optimization steps, but I expected 1k steps because I am using 8 TPU cores...).

---

Am I doing something wrong ? **How can I train on 8 TPU cores, with a batch size of 1 for each core ?**

### PR DESCRIPTION
**Pull request by jplu**
This PR is to make the TF trainer fully compliant with the TPUs.

Should fix #5042, #4996 and #4994.

### CODE REVIEW

File: examples/token-classification/run_tf_ner.py [hunk 1]
```diff
@@ -203,8 +203,8 @@ def compute_metrics(p: EvalPrediction) -> Dict:
     trainer = TFTrainer(
         model=model,
         args=training_args,
-        train_dataset=train_dataset.get_dataset() if train_dataset else None,
-        eval_dataset=eval_dataset.get_dataset() if eval_dataset else None,
+        train_sized_dataset=(train_dataset.get_dataset(), len(train_dataset)) if train_dataset else (None, None),
```
Comment (Colanim) [inline examples/token-classification/run_tf_ner.py:206]: Why should the user provide the size of dataset by himself ?

Wasn't it better when it was done automatically ? Also, it means user needs to take care of batch size in order to compute the right dataset size, am I right ?

**Comment by jplu** [comment]
@Colanim Can you summarize your findings here? It would be better suited than going across multiple issues.

**Comment by Colanim** [comment]
@jplu Sure !

1. Optimizer was passed as arguments to `_step` function, but we can pass only Tensors (issue #4994). This is fixed by ef02be8.

2. Optimization steps were not computed correctly : it was ignoring gradient accumulation (issue #5072). This is fixed by a897e09.

3. I'm still having error as described in #4996, but I didn't figure the reason yet. Maybe it's my model that has a problem, not `TFTrainer`.

File: src/transformers/trainer_tf.py [hunk 2]
```diff
@@ -84,10 +84,15 @@ def get_train_tfdataset(self) -> tf.data.Dataset:
...
+        batch_size = self.args.train_batch_size
+
+        if self.args.is_tpu_available():
+            batch_size *= self.args.gradient_accumulation_steps
+
         ds = (
             self.train_dataset.cache()
             .shuffle(self.num_train_examples)
-            .batch(self.args.train_batch_size, drop_remainder=self.args.dataloader_drop_last)
+            .batch(batch_size, drop_remainder=self.args.dataloader_drop_last)
```
Comment (Colanim) [inline src/transformers/trainer_tf.py:95]: Is this right ?

Batching with the whole `self.args.train_batch_size* self.args.gradient_accumulation_steps` ? I think we still want to batch on `self.args.train_batch_size`, so each training steps forward / backward only 1 batch.

Iteration on dataset is made here : https://github.com/huggingface/transformers/blob/efeb75b8054cc299698cf8bc09f395ada2660745/src/transformers/trainer_tf.py#L381 in `_accumulate_next_gradients`. 

And as we can see here : https://github.com/huggingface/transformers/blob/efeb75b8054cc299698cf8bc09f395ada2660745/src/transformers/trainer_tf.py#L354-L357 gradients are accumulated after going through `self.args.gradient_accumulation_steps` steps.

As far as I understand, we should do the batch/unbatch trick, as described in #5072 :
```python
ds = (
    self.train_dataset.cache()
    .shuffle(self.num_train_examples)
    .batch(self.args.train_batch_size * self.args.gradient_accumulation_steps, drop_remainder=self.args.dataloader_drop_last)
    .unbatch()
    .batch(self.args.train_batch_size, drop_remainder=self.args.dataloader_drop_last)
    .prefetch(tf.data.experimental.AUTOTUNE)
)
```

File: src/transformers/trainer_tf.py [hunk 3]
```diff
@@ -118,18 +129,14 @@ def get_optimizers(
...
-            self.args.warmup_steps,
-            adam_epsilon=self.args.adam_epsilon,
-            weight_decay_rate=self.args.weight_decay,
-        )
-
-        return optimizer, scheduler
+        if self.optimizers is None:
+            self.optimizers = create_optimizer(
+                self.args.learning_rate,
+                self.train_steps,
```
Comment (Colanim) [inline src/transformers/trainer_tf.py:135]: Isn't it supposed to be the total number of train steps during the training, not only 1 epoch ? i.e. `(self.num_train_examples / self.args.train_batch_size) * self.args.`

File: examples/token-classification/run_tf_ner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jplu) [reply examples/token-classification/run_tf_ner.py:206]: No, don't need to know the batch size or whatever else, just the number of examples. Anyway, this is done just because I got some errors on running the current way to compute the size on TPUs, because it cannot compute Python functions such as lambda.

This is also not definitive.

File: src/transformers/trainer_tf.py (hunk 2 above)
Comment (jplu) [reply src/transformers/trainer_tf.py:95]: I'm still not sure but ok, I trust you ^^ Will change this.

File: src/transformers/trainer_tf.py (hunk 3 above)
Comment (jplu) [reply src/transformers/trainer_tf.py:135]: This is fixed in another PR, see #5051.

**Comment by jplu** [comment]
Thanks for having sum up the issues here.

> * [ ]  Computation of number of steps should take into account `dataloader_drop_last` (mentioned in #5072), and use `math.floor` in that case (instead of `math.ceil`). Not fixed yet.

This is fixed in PR #5051.

> * [ ]  I'm still having error as described in #4996, but I didn't figure the reason yet. Maybe it's my model that has a problem, not `TFTrainer`.

This is what I'm currently looking for, but still not figuring out why :/

Which TPUs are you using, over ctpu, Cloud AI or Colab?

File: examples/token-classification/run_tf_ner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jplu) [reply examples/token-classification/run_tf_ner.py:206]: Did the original solution to compute the dataset size works on your side with TPUs?

[the remark this answers is not in the document]
Comment (Colanim) [reply examples/token-classification/run_tf_ner.py:206]: Yes, the original solution (`self.train_dataset.reduce(tf.constant(0), lambda x, _: x + 1).numpy()`) worked on my side. I'm using TF2.2, with TPU on GCP.

File: src/transformers/trainer_tf.py (hunk 3 above)
Comment (Colanim) [reply src/transformers/trainer_tf.py:140]: So many PR x)

File: src/transformers/trainer_tf.py (hunk 2 above)
Comment (Colanim) [reply src/transformers/trainer_tf.py]: Maybe we need the opinion of someone else :)

**Comment by Colanim** [comment]
> Which TPUs are you using, over ctpu, Cloud AI or Colab?

I'm currently using ctpu, but I could see similar issue when using Colab.

File: src/transformers/trainer_tf.py (hunk 2 above)
Comment (jplu) [reply src/transformers/trainer_tf.py]: I just found this https://www.kaggle.com/yihdarshieh/tpu-gradient-accumulation but the process is a bit different.

File: examples/token-classification/run_tf_ner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jplu) [reply examples/token-classification/run_tf_ner.py:206]: Ok, good to know then, it works for TF >= 2.2 only.

**Comment by Colanim** [comment]
For me, this is the problem : https://github.com/huggingface/transformers/blob/5f721ad6e48c9d846de25c3fefa0e50a306cbf10/src/transformers/trainer_tf.py#L388-L389

Somehow the Exception is not catched on TPU, which crash the training. Using `max_steps` argument instead of `num_train_epochs` fix the problem because we repeat the dataset, and therefore we never have an out of range error.

In the eval code, since we iterate the dataset without repeat, it's causing the same problem.

**Comment by jplu** [comment]
Can I see all your logs output during the training?

File: examples/token-classification/run_tf_ner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jplu) [reply examples/token-classification/run_tf_ner.py:206]: That's weird. I tried with ctpu with TF 2.2 and I get the following error:

```
2020-06-19 12:33:47.347085: W tensorflow/core/distributed_runtime/eager/remote_tensor_handle_data.cc:76] Unable to destroy remote tensor handles. If you are running a tf.function, it usually indicates some op in the graph gets an error: No registered 'PyFunc' OpKernel for 'CPU' devices compatible with node {{node PyFunc}}
	.  Registered:  <no registered kernels>

	 [[PyFunc]]
Traceback (most recent call last):
  File "run_tf_ner.py", line 281, in <module>
    main()
  File "run_tf_ner.py", line 213, in main
    trainer.train()
  File "/home/jplu/transformers/src/transformers/trainer_tf.py", line 276, in train
    train_ds = self.get_train_tfdataset()
  File "/home/jplu/transformers/src/transformers/trainer_tf.py", line 73, in get_train_tfdataset
    self.num_train_examples = self.train_dataset.reduce(tf.constant(0), lambda x, _: x + 1).numpy()
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/ops.py", line 961, in numpy
    maybe_arr = self._numpy()  # pylint: disable=protected-access
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/ops.py", line 929, in _numpy
    six.raise_from(core._status_to_exception(e.code, e.message), None)
  File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.NotFoundError: No registered 'PyFunc' OpKernel for 'CPU' devices compatible with node {{node PyFunc}}
	.  Registered:  <no registered kernels>

	 [[PyFunc]]
```

**Comment by jplu** [comment]
What is the command line you are using to create your TPU and run the process? In order to be aligned with the same errors :)

**Comment by Colanim** [comment]
@jplu finally I was wrong : even when using `max_steps` I'm having the problem.

It always happen at the beginning of the end of first epoch (for both validation and training). I think it's due to how the dataset is iterated. 

According to [this link](https://www.kaggle.com/mgornergoogle/custom-training-loop-with-100-flowers-on-tpu), we should have a single iterator for the whole training procedure (iterator should not be reset after each epoch), and always `repeat` the dataset indefinitely. Even for validation dataset.

Unfortunately I'm having trouble with GCP right now, I can't try things on my end..

**Comment by jplu** [comment]
Ok, can you share your command lines here please. Because even the size computation cannot be run on TPU with TF 2.2.

File: src/transformers/trainer_tf.py [hunk 4]
```diff
@@ -70,20 +72,34 @@ def __init__(
...
+        else:
+            ds = (
+                self.train_dataset.cache()
+                .shuffle(self.num_train_examples)
+                .batch(self.args.train_batch_size, drop_remainder=self.args.dataloader_drop_last)
+                .prefetch(tf.data.experimental.AUTOTUNE)
+            )
 
         if self.args.max_steps > 0:
             self.train_dataset = self.train_dataset.repeat(-1)
```
Comment (Colanim) [inline src/transformers/trainer_tf.py:105]: @jplu Aha ! This is the error ! I found it !!

So as I mentioned in a previous comment, apparently we can't catch `OutOfRange` exception on TPU, which means we can't iterate the dataset epoch by epoch. Instead we should **ALWAYS** repeat the dataset, and iterate a specific number of steps. Same for validation / test dataset, as mentioned in the Kaggle notebook I linked previously. So there is quite a lot of change to make it work on TPU...

Anyway this line is wrong, it should be :
```python
ds = ds.repeat()
```

**Review by astariul — CHANGES_REQUESTED** [review]

File: src/transformers/trainer_tf.py (hunk 4 above)
Comment (jplu) [reply src/transformers/trainer_tf.py:105]: Repeat is needed only if you run with `max_steps` > 1. But for the same reason `.repeat()` should be applied over the validation dataset indeed.

Comment (Colanim) [reply src/transformers/trainer_tf.py:105]: In the case where the user specify `num_train_epochs` (`max_steps` = 0), the code rely on this try-catch to not iterate too far :

https://github.com/huggingface/transformers/blob/514410407019e9d20de7c2555595bab145c39ac0/src/transformers/trainer_tf.py#L385-L389

In my case, this try-catch is not working on TPU. So I need to rely on the behavior of `max_steps` > 1.

---

In the Kaggle notebook linked above, they mentioned that recreating a new iterator every epoch slow the training because TPU needs to re-compile the graph at the beginning of every epoch. Maybe we want to keep providing the `num_train_epochs` / `max_steps` behavior, but under the hood use only 1 way to iterate : `repeat()` with specific number of steps. 

What do you think ?

Comment (jplu) [reply src/transformers/trainer_tf.py:105]: You are right. What I have more in mind is: if someone wants to use TPUs he has to fullfil the `max_steps` parameter and keep like it is for GPUs/CPU.

The why of this try/catch is because it is more optimized for dataset iteration for CPU/GPUs.

Comment (jplu) [reply src/transformers/trainer_tf.py:105]: Later I will rethink the full process in order to be able to better handle the three cases (CPU/GPUs/TPUs).
